### ISSUE TITLE
value_or(None) is Type-Unsafe - Add unwrap() and expect() methods

### ISSUE THREAD

**Issue by mikelane** [mikelane/valid8r#271]
## Summary

After checking `is_success()`, type checkers still think `value_or(None)` might return `None`. This forces unnecessary type narrowing or `# type: ignore` comments throughout user code.

## Problem

**Current behavior:**
```python
result = parse_int("42")
if result.is_success():
    value = result.value_or(None)  # Type: int | None (wrong!)
    # Type checker still thinks this could be None
    print(value + 1)  # Error: None + 1
```

**Expected behavior:**
```python
result = parse_int("42")
if result.is_success():
    value = result.unwrap()  # Type: int (correct!)
    print(value + 1)  # Works
```

## User Impact

- Constant type checker warnings in valid8r code
- Users add `# type: ignore` comments everywhere
- IDE type hints are wrong after `is_success()` checks
- Pattern matching is the only type-safe extraction method

## Acceptance Criteria

```gherkin
Feature: Type-safe value extraction from Maybe

  Scenario: Extract value with unwrap when Success
    Given I have a Success containing the integer 42
    When I call unwrap() on the result
    Then I receive the integer 42
    And the return type is int, not int | None

  Scenario: Extract value with unwrap when Failure
    Given I have a Failure containing "invalid input"
    When I call unwrap() on the result
    Then an UnwrapError is raised
    And the error message contains "invalid input"

  Scenario: Extract value with expect and custom message
    Given I have a Failure containing "parse error"
    When I call expect("Age must be valid") on the result
    Then an UnwrapError is raised
    And the error message contains "Age must be valid"

  Scenario: Type checker narrows type after is_success check
    Given I have a Maybe[int] result
    When I check is_success() returns True
    And I call unwrap() inside the if block
    Then the type checker infers the return type as int
    And no type errors are reported

  Scenario: Extract error with unwrap_err when Failure
    Given I have a Failure containing "validation failed"
    When I call unwrap_err() on the result
    Then I receive the string "validation failed"
    And the return type is str, not str | None

  Scenario: Extract error with unwrap_err when Success
    Given I have a Success containing the integer 42
    When I call unwrap_err() on the result
    Then an UnwrapError is raised
    And the error message indicates this was a Success
```

### New Methods to Add

```python
class Maybe(Generic[T]):
    def unwrap(self) -> T:
        """Extract value or raise UnwrapError if Failure."""
        if self.is_success():
            return self._value  # type: ignore
        raise UnwrapError(f"Called unwrap() on Failure: {self._error}")
    
    def expect(self, msg: str) -> T:
        """Extract value or raise UnwrapError with custom message."""
        if self.is_success():
            return self._value  # type: ignore
        raise UnwrapError(msg)
    
    def unwrap_err(self) -> str:
        """Extract error or raise UnwrapError if Success."""
        if self.is_failure():
            return self._error  # type: ignore
        raise UnwrapError("Called unwrap_err() on Success")
```

### TypeGuard for is_success()

Consider adding a TypeGuard to enable proper narrowing:

```python
from typing import TypeGuard

class Maybe(Generic[T]):
    def is_success(self) -> TypeGuard[Success[T]]:
        return self._is_success
```

**Note**: This may require splitting `Maybe` into `Success` and `Failure` subclasses or using `@overload`.

### New Exception Class

```python
class UnwrapError(Exception):
    """Raised when unwrap() is called on wrong variant."""
    pass
```

## Dependencies

**This issue unblocks:**
- #TBD (`and_then` alias for `bind`)
- #TBD (Error collection methods)
- #TBD (`Maybe.from_optional()` helper)
- #TBD (`Maybe.to_optional()` helper)

## Definition of Done

- [ ] `unwrap()` method added to Maybe
- [ ] `expect(msg)` method added to Maybe
- [ ] `unwrap_err()` method added to Maybe
- [ ] `UnwrapError` exception class created
- [ ] TypeGuard considered for `is_success()` / `is_failure()`
- [ ] All methods have correct type annotations
- [ ] Unit tests for all new methods
- [ ] Documentation updated with examples
- [ ] mypy passes with strict settings

## Size Estimate

**Medium (M)**: ~1-2 days (type system complexity)

## Related

- **Epic**: #269 (Developer Experience Improvements)
- **Priority**: P0 (Critical - type safety affects all users)
- **Blocks**: Multiple P1 and P3 issues
